### ISSUE TITLE
Where is the benchmark about 90 times faster than HF transformers?

### ISSUE THREAD

**Issue by hyunwoongko** [legacyai/tf-transformers#39]
You said this library is 90 times faster than HF transformers, but there is no benchmark about it.
https://github.com/legacyai/tf-transformers/tree/main/benchmarks

**Comment by legacyai** [comment]
Hi @hyunwoongko . Thanks for opening the first issue.
You can find it here .

I tested it on V100 GPU machine. 

Code to test: https://github.com/legacyai/tf-transformers/tree/main/benchmarks/gpt2
Website docs : https://legacyai.github.io/tf-transformers/build/html/benchmarks/gpt2.html

<img width="1076" alt="Screenshot 2022-04-09 at 11 26 28 AM" src="https://user-images.githubusercontent.com/67182839/162558719-c2ed110f-ae49-48ba-aa78-dd350e3aaeab.png">

**Comment by hyunwoongko** [comment]
Yeah. But 31 minute is not 90 times faster than 83 minute. no?
Is there anything I misunderstood?

**Comment by legacyai** [comment]
Yes I agree. Exact relative difference from the above benchmark is ( 62x ) times.
But, trust me, in many cases I have even seen 120x times speedup.
But for that other factors comes into picture, variable batch decoding, eos token, sequence length etc. The number ( 90x ) was something I approximated between 62x and 120x. 

I have. tutorial ( https://legacyai.github.io/tf-transformers/build/html/tutorials/7_gpt2_question_answering_squad.html ) here, you can experience similar speedup.

**Comment by hyunwoongko** [comment]
nobody thinks `31 minutes` is 90 times faster than `83 minutes`. it's 2.67x faster.

**Comment by hyunwoongko** [comment]
In addition, relative difference can't express as `times`. 
For example, relative difference of 1 and 2 is 66.7%
so, 2 is 66 times larger than 1? it's not. 
2 is 2 times larger than 1.
66 is 66 times larger than 1. right?

**Comment by hyunwoongko** [comment]
I admire you for developing such a wonderful library. However, skewing the benchmarks is a problem, and it makes users untrustworthy of this library. I recommend that you correct any incorrect expressions in the documentation.

**Comment by legacyai** [comment]
Agreed . I always calculated ```relative difference``` to measure, and apologies for wrong usage of word ```times```  as such. I always had this in my mind and updated the docs with new information. From ```90 times``` to ```90 %``` improvement.

**Comment by legacyai** [comment]
> I admire you for developing such a wonderful library. However, skewing the benchmarks is a problem, and it makes users untrustworthy of this library. I recommend that you correct any incorrect expressions in the documentation.

I don't think, it was supposed to ```make users untrustworthy```.  If thats the case, I wouldn't have provided all the scripts and code to reproduce benchmark. But, yes It was mistake from my end from wrong usage of ```times```.

**Comment by legacyai** [comment]
> 

If I am not wrong ```relative difference``` is a valid measure to calculate the speedup right. Yes, it can't be expressed as times, my bad. But I believe the benchmarking is very much valid.

**Comment by aveni** [comment]
Just want to second what @hyunwoongko said, the difference from 83min -> 31min should be expressed as either:

* (31/83) = 0.37 -> decoding takes 37% of the time as it did before
* or (83/31) = 2.67 -> decoding is performed 2.67x faster than before

The `relative difference = |a - b| / mean(a, b)` does not make any sense to me as a measure for comparing minutes or throughputs. Unlike the two meaures above, `relative difference` does not give me a simple way to predict how much faster my own workload would get if I used your library. I would rather not have to solve `b = a * (2 - relative_difference) / (2 + relative_difference)` in my head.

**Comment by legacyai** [comment]
@aveni  - This is tutorial from huggingface https://huggingface.co/blog/tf-serving . 

<img width="803" alt="image" src="https://user-images.githubusercontent.com/67182839/162596312-3181115f-8096-4dc8-b26f-0bc214a48730.png">

If you see they expressed the speedup in terms of relative difference. I checked internet for ```how to measure speedup``` and ```relative difference``` seems to be a valid one. But I could be wrong too, I followed it because it was used in the above blog officially and it gives as sense of ```percentage improvement``` also.
